substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/elf/dwarf/DwarfSections.java

```diff
@@ -0,0 +1,2344 @@
+        }
+        String filePath = file.substring(0, pathLength);
+        DirEntry dirEntry = dirsIndex.get(filePath);
+        if (dirEntry == null) {
+            dirEntry = new DirEntry(stringTable.uniqueString(filePath));
+            dirsIndex.put(filePath, dirEntry);
+            dirs.add(dirEntry);
+        }
+        return dirEntry;
+    }
+
+    /**
+     * class from which all DWARF debug section
+     * inherit providing common behaviours.
+     */
+    // shared implementation methods to manage content creation
+    public abstract class DwarfSectionImpl extends BasicProgbitsSectionImpl {
+        public boolean debug = false;
+        public long debugTextBase = 0;
+        public long debugAddress = 0;
+        public int debugBase = 0;
+
+        public DwarfSectionImpl() {
+        }
+
+        /**
+         * creates the target byte[] array used to define the section
+         * contents.
+         *
+         * the main task of this method is to precompute the
+         * size of the debug section. given the complexity of the
+         * data layouts that invariably requires performing a dummy
+         * write of the contents, inserting bytes into a small,
+         * scratch buffer only when absolutely necessary. subclasses
+         * may also cache some information for use when writing the
+         * contents.
+         */
+        public abstract void createContent();
+
+        /**
+         * populates the byte[] array used to contain the section
+         * contents.
+         *
+         * in most cases this task reruns the operations performed
+         * under createContent but this time actually writing data
+         * to the target byte[].
+         */
+        public abstract void writeContent();
+
+        @Override
+        public boolean isLoadable() {
+            // even though we're a progbits section impl we're not actually loadable
+            return false;
+        }
+
+        public void checkDebug(int pos) {
+            // if the env var relevant to this element
+            // type is set then switch on debugging
+            String name = getSectionName();
+            String envVarName = "DWARF_" + name.substring(1).toUpperCase();
+            if (System.getenv(envVarName) != null) {
+                debug = true;
+                debugBase = pos;
+                debugAddress = debugTextBase;
+            }
+        }
+
+        protected void debug(String format, Object... args) {
+            if (debug) {
+                System.out.format(format, args);
+            }
+        }
+
+        // base level put methods that assume a non-null buffer
+
+        public int putByte(byte b, byte[] buffer, int p) {
+            int pos = p;
+            buffer[pos++] = b;
+            return pos;
+        }
+
+        public int putShort(short s, byte[] buffer, int p) {
+            int pos = p;
+            buffer[pos++] = (byte) (s & 0xff);
+            buffer[pos++] = (byte) ((s >> 8) & 0xff);
```

Comment:
Does DWARF have a fixed endianness, or should it be using the ELFDATA value to determine this?

Reply:
Hmm, the DWARF standard didn't really make that clear but I talked to one of our debug experts (Mark Wielaard) and he confirmed that most numeric values need to follow target arch endianness. The special cases are ULEB and SLEB encoded numerics. So, I will update the code accordingly.

Reply:
I have pushed a fix that addresses the byte order issue.